pkg/controller/agent/pod.go

```diff
@@ -323,3 +320,111 @@ func certificatesDir(association commonv1.Association) string {
 		ref.Name,
 	)
 }
+
+func getFleetModeEnvVars(agent agentv1alpha1.Agent, client k8s.Client) ([]corev1.EnvVar, error) {
+	result := []corev1.EnvVar{}
+
+	for _, f := range []func(agentv1alpha1.Agent, k8s.Client) ([]corev1.EnvVar, error){
+		getFleetSetupKibanaEnvVars,
+		getFleetSetupFleetEnvVars,
+		getFleetSetupFleetServerEnvVars,
+	} {
+		envVars, err := f(agent, client)
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, envVars...)
+	}
+
+	return result, nil
+}
+
+func getFleetSetupKibanaEnvVars(agent agentv1alpha1.Agent, client k8s.Client) ([]corev1.EnvVar, error) {
+	if agent.Spec.KibanaRef.IsDefined() {
+		kbConnectionSettings, err := extractConnectionSettings(agent, client, commonv1.KibanaAssociationType)
+		if err != nil {
+			return nil, err
+		}
+
+		envVars := []corev1.EnvVar{
+			{Name: "KIBANA_FLEET_HOST", Value: kbConnectionSettings.host},
+			{Name: "KIBANA_FLEET_USERNAME", Value: kbConnectionSettings.username},
+			{Name: "KIBANA_FLEET_PASSWORD", Value: kbConnectionSettings.password},
+			{Name: "KIBANA_FLEET_SETUP", Value: strconv.FormatBool(agent.Spec.KibanaRef.IsDefined())},
+		}
+
+		// don't set ca key if ca is not available
+		if kbConnectionSettings.ca != "" {
+			envVars = append(envVars, corev1.EnvVar{Name: "KIBANA_FLEET_CA", Value: kbConnectionSettings.ca})
+		}
+
+		return envVars, nil
+	}
+
+	return []corev1.EnvVar{}, nil
+}
+
+func getFleetSetupFleetEnvVars(agent agentv1alpha1.Agent, client k8s.Client) ([]corev1.EnvVar, error) {
+	fleetCfg := []corev1.EnvVar{}
+
+	if agent.Spec.KibanaRef.IsDefined() {
+		fleetCfg = append(fleetCfg, corev1.EnvVar{Name: "FLEET_ENROLL", Value: "true"})
+	}
+
+	if agent.Spec.FleetServerEnabled {
+		fleetURL, err := association.ServiceURL(
+			client,
+			types.NamespacedName{Namespace: agent.Namespace, Name: HTTPServiceName(agent.Name)},
+			agent.Spec.HTTP.Protocol(),
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		fleetCfg = append(fleetCfg, corev1.EnvVar{Name: "FLEET_CA", Value: path.Join(FleetCertsMountPath, certificates.CAFileName)})
+		fleetCfg = append(fleetCfg, corev1.EnvVar{Name: "FLEET_URL", Value: fleetURL})
+	} else if agent.Spec.FleetServerRef.IsDefined() {
```

Comment:
Maybe a comment would help the casual reader remember the different modes we have: 

* Standalone agent
* Agent acting as Fleet server
* Agent enrolled in Fleet (I think that is what `FleetServerRef` indicates